e2e/cypress/integration/signin_authentication/authentication_spec.js

```diff
@@ -0,0 +1,283 @@
+        cy.get('#logout').should('be.visible').click();
+
+        // # Login as user B and switch to a different team and channel
+        cy.visit('/login');
+        fillCredentialsForUser(testUser2);
+
+        cy.visit(`/${testTeam.name}/channels/off-topic`).wait(timeouts.ONE_SEC);
+
+        // # Logout
+        cy.get('#sidebarHeaderDropdownButton').should('be.visible').click();
+        cy.get('#logout').should('be.visible').click();
+
+        // # Login as user B again, observe you're viewing the team/channel you switched to in step 3
+        cy.visit('/login');
+        fillCredentialsForUser(testUser2);
+
+        cy.url().should('include', `/${testTeam.name}/channels/off-topic`);
+    });
+
+    it('MM-T419 Desktop session expires when the focus is on the tab', () => {
+        cy.visit('/login');
+        fillCredentialsForUser(testUser);
+
+        // # From Account Settings, set your desktop notifications to Never (if not already done)
+        // Click hamburger main menu.
+        cy.get('#sidebarHeaderDropdownButton').click();
+
+        // Click "Account settings"
+        cy.findByText('Account Settings').should('be.visible').click();
+
+        // Check that the "Account Settings" modal was opened.
+        cy.get('#accountSettingsModal').should('exist').within(() => {
+            // Click "Notifications"
+            cy.findByText('Notifications').should('be.visible').click();
+
+            // Click "Desktop"
+            cy.findByText('Desktop Notifications').should('be.visible').click();
+
+            // # Set your desktop notifications to Never
+            cy.get('#desktopNotificationNever').check();
+
+            // Click "Save"
+            cy.findByText('Save').scrollIntoView().should('be.visible').click();
+
+            // Close the modal.
+            cy.get('#accountSettingsHeader').find('button').should('be.visible').click();
+        });
+
+        cy.visit(`/${testTeam.name}/channels/town-square`).wait(timeouts.ONE_SEC);
+
+        // Visit the town-square with the notification API stubbed.
+        spyNotificationAs('withNotification', 'granted');
+
+        // # From a separate browser session, login to the same server as the same user
+        // # Click the hamburger menu and select Account Settings ➜ Security
+        // # Click "View and Logout of Active Sessions", then find and close the session created in step 1
+        // Since we are testing this on browser, we can revoke sessions with admin user.
+        const sysadmin = getAdminAccount();
+        cy.externalRequest({user: sysadmin, method: 'post', path: `users/${testUser.id}/sessions/revoke/all`}).wait(timeouts.HALF_MIN);
+
+        // # Go back and view the original session app/browser, and wait until you see a desktop notification (may take up to a minute)
+        // * Desktop notification is sent (may take up to 1 min)
+        //cy.get('@withNotification').should('have.been.calledOnce');
```

Comment:
Might need some help here, it does not receive the notification even though it does when I test this manually. Probably I'm missing something here.

Reply:
I think the stub to Notification gets wiped out on redirect to `/login` page (new page load). I got it working by directly stubbing on App event (`'window:before:load'`). Note: works consistently when this test is on a separate spec file.
```javascript
    it('MM-T419 Desktop session expires when the focus is on the tab', () => {
        Cypress.on('window:before:load', (win) => {
            function Notification(title, opts) {
                this.title = title;
                this.opts = opts;
            }
    
            Notification.requestPermission = () => 'granted';
            Notification.close = () => true;
    
            win.Notification = Notification;
    
            cy.stub(win, 'Notification').as('withNotification');
        });

        cy.visit('/login');
        fillCredentialsForUser(testUser);

        // # From Account Settings, set your desktop notifications to Never (if not already done)
        // Click hamburger main menu.
        cy.get('#sidebarHeaderDropdownButton').click();

        // Click "Account settings"
        cy.findByText('Account Settings').should('be.visible').click();

        // Check that the "Account Settings" modal was opened.
        cy.get('#accountSettingsModal').should('exist').within(() => {
            // Click "Notifications"
            cy.findByText('Notifications').should('be.visible').click();

            // Click "Desktop"
            cy.findByText('Desktop Notifications').should('be.visible').click();

            // # Set your desktop notifications to Never
            cy.get('#desktopNotificationNever').check();

            // Click "Save"
            cy.findByText('Save').scrollIntoView().should('be.visible').click();

            // Close the modal.
            cy.get('#accountSettingsHeader').find('button').should('be.visible').click();
        });

        // # Visit the town-square with the notification API stubbed.
        cy.visit(`/${testTeam.name}/channels/town-square`).wait(timeouts.ONE_SEC);

        // # From a separate browser session, login to the same server as the same user
        // # Click the hamburger menu and select Account Settings ➜ Security
        // # Click "View and Logout of Active Sessions", then find and close the session created in step 1
        // Since we are testing this on browser, we can revoke sessions with admin user.
        const sysadmin = getAdminAccount();
        cy.externalRequest({user: sysadmin, method: 'post', path: `users/${testUser.id}/sessions/revoke/all`}).wait(timeouts.ONE_SEC);

        // # Go back and view the original session app/browser, and wait until you see a desktop notification (may take up to a minute)
        // * Desktop notification is sent (may take up to 1 min)
        cy.get('@withNotification').should('have.been.calledOnce').and('have.been.calledWithMatch', 'Mattermost', ({body}) => {
            const expected = 'Session Expired: Please sign in to continue receiving notifications.';
            expect(body, `Notification body: "${body}" should match: "${expected}"`).to.equal(expected);
            return true;
        });

        // * Login page shows a message above the login box that the session has expired.
        cy.get('#login_section .alert-warning').should('contain.text', 'Your session has expired. Please log in again.');
    });
```

Reply:
Thanks a lot @saturninoabril I would never guessed this 😅